cpp/src/arrow/util/future.cc

```diff
@@ -231,26 +232,68 @@ class ConcreteFutureImpl : public FutureImpl {
 
   void DoMarkFailed() { DoMarkFinishedOrFailed(FutureState::FAILURE); }
 
-  void AddCallback(Callback callback) {
+  void CheckOptions(const CallbackOptions& opts) {
+    if (opts.should_schedule != ShouldSchedule::NEVER) {
+      DCHECK_NE(opts.executor, NULL)
+          << "An executor must be specified when adding a callback that might schedule";
+    }
+  }
+
+  void AddCallback(Callback callback, CallbackOptions opts) {
+    CheckOptions(opts);
     std::unique_lock<std::mutex> lock(mutex_);
+    CallbackRecord callback_record{std::move(callback), opts};
```

Comment:
Why doesn't `AddCallback` directly take a `CallbackRecord`?

Reply:
I'm only combining the two out of convenience for storage.  It seems like an implementation detail that doesn't need to be exposed.  Also, if I take a `CallbackRecord` directly here it will be asymmetric with `TryAddCallback` which can't take a `CallbackRecord` (unless we want the factory to return a callback record which exposes all of this outside the file as well).